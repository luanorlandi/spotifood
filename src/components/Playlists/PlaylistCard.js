import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';

import './PlaylistCard.scss';

const PlaylistCard = ({ playlist }) => {
  const { name, description } = playlist;
  const imageUrl = get(playlist, 'images[0].url');
  const spotifyLink = get(playlist, 'external_urls.spotify');

  return (
    <a href={spotifyLink} className="playlist-card">
      <img
        className="playlist-card__image"
        src={imageUrl}
        alt="Capa da playlist"
      />
      <div>
        <div className="playlist-card__title">{name}</div>
        <div>{description}</div>
      </div>
    </a>
  );
};

PlaylistCard.propTypes = {
  playlist: PropTypes.shape({
    name: PropTypes.string,
    description: PropTypes.string,
    images: PropTypes.array,
    external_urls: PropTypes.object,
  }).isRequired,
};

export default PlaylistCard;
