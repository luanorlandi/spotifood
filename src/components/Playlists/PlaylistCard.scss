@import '~bulma/sass/utilities/_all.sass';

.playlist-card {
  padding: 16px;
  background-color: #252525;
  color: #b3b3b3;
  transition: 200ms;

  &:hover {
    color: #b3b3b3;
    background-color: #333333;
  }

  @include touch {
    text-align: center;
  }
}

.playlist-card__image {
  width: 300px;
  height: 268px;
}

.playlist-card__title {
  color: #fff;
  font-size: 1.1rem;
  font-weight: 700;
}