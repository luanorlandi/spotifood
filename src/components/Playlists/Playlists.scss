@import '~bulma/sass/utilities/_all.sass';

.playlists {
  width: 100%;
  padding: 20px;
}

.playlists__cards {
  display: grid;
  grid-template-columns: repeat(auto-fill, 300px);
  grid-auto-rows: auto;
  grid-gap: 20px;

  @include touch {
    justify-content: center;
  }
}
