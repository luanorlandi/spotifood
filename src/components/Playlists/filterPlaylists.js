const filterPlaylists = (playlists, filter) => {
  const lowerCaseFilter = filter.toLowerCase();

  return playlists.filter((playlist) => {
    const playlistName = playlist.name.toLowerCase();
    return playlistName.includes(lowerCaseFilter);
  });
};

export default filterPlaylists;
