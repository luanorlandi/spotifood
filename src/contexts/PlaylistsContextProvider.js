import React, { useState } from 'react';
import PropTypes from 'prop-types';

import PlaylistsContext from './PlaylistsContext';

const PlaylistsContextProvider = ({ children, initialPlaylists }) => {
  const [playlists, setPlaylists] = useState(initialPlaylists);

  return (
    <PlaylistsContext.Provider value={{ playlists, setPlaylists }}>
      {children}
    </PlaylistsContext.Provider>
  );
};

PlaylistsContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
  initialPlaylists: PropTypes.array,
};

PlaylistsContextProvider.defaultProps = {
  initialPlaylists: null,
};

export default PlaylistsContextProvider;
