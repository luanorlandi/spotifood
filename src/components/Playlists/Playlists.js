import React, { useContext, useState, useEffect } from 'react';
import { useFormik } from 'formik';

import PlaylistsContext from '../../contexts/PlaylistsContext';
import Input from '../Form/Input/Input';
import PlaylistCard from './PlaylistCard';
import filterPlaylists from './filterPlaylists';
import './Playlists.scss';

const Playlists = () => {
  const { playlists } = useContext(PlaylistsContext);
  const [filteredPlaylists, setFilteredPlaylists] = useState([]);
  const formik = useFormik({
    initialValues: {
      name: '',
    },
  });

  useEffect(() => {
    const hasfilter = !!formik.values.name;
    if (playlists && hasfilter) {
      const newFilteredPlaylists = filterPlaylists(playlists, formik.values.name);
      setFilteredPlaylists(newFilteredPlaylists);
    } else {
      setFilteredPlaylists(playlists);
    }
  }, [playlists, formik.values.name]);

  return (
    <div className="playlists">
      <Input
        formik={formik}
        field={{ id: 'name' }}
        placeholder="Filtrar pelo nome da playlist"
        styleType="white"
      />
      <div className="playlists__cards">
        {filteredPlaylists && filteredPlaylists.map((playlist) => (
          <PlaylistCard key={playlist.id} playlist={playlist} />
        ))}
      </div>
    </div>
  );
};

export default Playlists;
