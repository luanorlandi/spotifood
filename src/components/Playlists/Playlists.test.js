import React from 'react';
import { render, fireEvent, waitForElementToBeRemoved } from '@testing-library/react';

import Playlists from './Playlists';
import PlaylistsContextProvider from '../../contexts/PlaylistsContextProvider';

const playlists = [{
  id: '37i9dQZF1DXbYM3nMM0oPk',
  name: 'Mega Hit Mix',
  description: 'A mega mix of 75 of your favorite songs from the last few years! Cover: Post Malone',
  images: [{
    url: 'https://i.scdn.co/image/ab67706f00000002002b3728f0d5c5c599b9b06a',
  }],
  external_urls: {
    spotify: 'https://open.spotify.com/playlist/37i9dQZF1DXbYM3nMM0oPk',
  },
}, {
  id: '37i9dQZF1DX4sWSpwq3LiO',
  name: 'Peaceful Piano',
  description: 'Relax and indulge with beautiful piano pieces',
  images: [{
    url: 'https://i.scdn.co/image/ab67706f00000002ca5a7517156021292e5663a6',
  }],
  external_urls: {
    spotify: 'https://open.spotify.com/playlist/37i9dQZF1DX4sWSpwq3LiO',
  },
}];

test('playlists match snapshot', () => {
  const { asFragment } = render((
    <PlaylistsContextProvider initialPlaylists={playlists}>
      <Playlists />
    </PlaylistsContextProvider>
  ));

  expect(asFragment()).toMatchSnapshot();
});

test('filter playlists by name', async () => {
  const { getByPlaceholderText, getByText } = render((
    <PlaylistsContextProvider initialPlaylists={playlists}>
      <Playlists />
    </PlaylistsContextProvider>
  ));

  const filterPromise = waitForElementToBeRemoved(() => getByText(playlists[1].name));

  fireEvent.change(getByPlaceholderText('Filtrar pelo nome da playlist'), {
    target: { value: playlists[0].name },
  });

  await filterPromise;

  expect(getByText(playlists[0].name)).toBeInTheDocument();
});
