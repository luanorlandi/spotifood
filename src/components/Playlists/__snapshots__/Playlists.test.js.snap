// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playlists match snapshot 1`] = `
<DocumentFragment>
  <div
    class="playlists"
  >
    <div
      class="field"
    >
      <label
        for="name"
      />
      <input
        class="input input--is-white"
        id="name"
        min="0"
        placeholder="Filtrar pelo nome da playlist"
        step="1"
        type="text"
        value=""
      />
    </div>
    <div
      class="playlists__cards"
    >
      <a
        class="playlist-card"
        href="https://open.spotify.com/playlist/37i9dQZF1DXbYM3nMM0oPk"
      >
        <img
          alt="Capa da playlist"
          class="playlist-card__image"
          src="https://i.scdn.co/image/ab67706f00000002002b3728f0d5c5c599b9b06a"
        />
        <div>
          <div
            class="playlist-card__title"
          >
            Mega Hit Mix
          </div>
          <div>
            A mega mix of 75 of your favorite songs from the last few years! Cover: Post Malone
          </div>
        </div>
      </a>
      <a
        class="playlist-card"
        href="https://open.spotify.com/playlist/37i9dQZF1DX4sWSpwq3LiO"
      >
        <img
          alt="Capa da playlist"
          class="playlist-card__image"
          src="https://i.scdn.co/image/ab67706f00000002ca5a7517156021292e5663a6"
        />
        <div>
          <div
            class="playlist-card__title"
          >
            Peaceful Piano
          </div>
          <div>
            Relax and indulge with beautiful piano pieces
          </div>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
